src/Tactic/Inline.agda

```diff
@@ -0,0 +1,279 @@
+  evens? []       = yes []
+  evens? (x ∷ xs) = mapDec (uncurry _∷_) uncons (even? x ×-dec evens? xs)
+  -}
+  _ = evens? (0 ∷ 2 ∷ []) ≡ yes (zero ∷ suc zero ∷ [])
+    ∋ refl
+
+  -- (3) works under module parameters
+  module _ (n m : ℕ) where
+    unquoteDecl ⟫evens? = inlineDecl ⟫evens? (quoteTerm (all? even?))
+    _ = ⟫evens? (0 ∷ 2 ∷ []) ≡ yes (zero ∷ suc zero ∷ [])
+      ∋ refl
+
+  module _ {A B : Set} (f : A → B) where
+    map' : List A → List B
+    map' [] = []
+    map' (x ∷ xs) = f x ∷ map' xs
+
+  unquoteDecl sucs' = inlineDecl sucs' (quoteTerm (map' {B = ℕ} suc))
+  _ = sucs' (0 ∷ 1 ∷ 2 ∷ 3 ∷ []) ≡ (1 ∷ 2 ∷ 3 ∷ 4 ∷ [])
+    ∋ refl
+
+  -- (4) works under mutual blocks
+  data Odd : ℕ → Set where
+    one : Odd 1
+    suc : ∀ {n} → Odd n → Odd (suc (suc n))
+
+  mutual
+    odd? : Decidable¹ Odd
+    odd? = λ where
+      0 → no λ ()
+      1 → yes one
+      (suc (suc n)) → mapDec suc (λ where (suc p) → p) (odd? n)
+
+    unquoteDecl mevens? = inlineDecl mevens? (quoteTerm (all? even?))
+    unquoteDecl modds?  = inlineDecl modds?  (quoteTerm (all? odd?))
+    _ = mevens? (0 ∷ 2 ∷ []) ≡ yes (zero ∷ suc zero ∷ [])
+      ∋ refl
+
+    -- [AGDA BUG] cannot use _∋_ (c.f. Agda issue #7028)
+    _ : modds? (1 ∷ 3 ∷ []) ≡ yes (one ∷ suc one ∷ [])
+    _ = refl
+
+  -- (5) works with `with`-statements (e.g. for specializing `mapMaybe`)
+  toEvenOdd : ℕ → ℕ {- even part -} × Maybe ℕ {- odd part -}
+  toEvenOdd n with even? n
+  ... | yes _ = n , nothing
+  ... | no  _ = pred n , just 1
+
+  toOdd : ℕ → Maybe ℕ
+  toOdd = proj₂ ∘ toEvenOdd
+
+  unquoteDecl toOdds = inlineDecl toOdds (quoteTerm (L.mapMaybe toOdd))
+  {-
+  toOdds : List ℕ → List ℕ
+  toOdds []       = []
+  toOdds (x ∷ xs) with toOdd x
+  -- ** [LIMITATION] does not recursively inline `with`-statements
+  -- ... | just y  = y ∷ toOdds xs
+  -- ... | nothing = toOdds xs
+  ... | just y  = y ∷ mapMaybe toOdd xs
+  ... | nothing = mapMaybe toOdd xs
+  -}
+  _ = toOdds (0 ∷ 1 ∷ 2 ∷ 3 ∷ []) ≡ (1 ∷ 1 ∷ [])
+    ∋ refl
+
+  -- (6) test for `MOf?`
+  open import Data.List.Relation.Unary.MOf using (MOf; mOf; MOf?)
+
+  unquoteDecl MOf-even? = inlineDecl MOf-even? (quoteTerm (MOf? even?))
+  {-
+  MOf?-even? : ∀ m xs → Dec (MOf m Even xs)
+  MOf?-even? zero    xs = yes done
+  MOf?-even? (suc m) [] = no λ where (mOf (_ ∷ _) len≡ () _)
+  MOf?-even? (suc m) (x ∷ xs) =
+    if even? x then
+      (λ {px} → mapDec (cons px) uncons (MOf?-even? m xs))
+    else
+      (λ {¬px} → mapDec skip (unskip ¬px) (MOf?-even? (suc m) xs))
+  -}
+
+  _ = MOf-even? 2 (0 ∷ 1 ∷ 2 ∷ 3 ∷ [])
+    ≡ yes (mOf (0 ∷ 2 ∷ []) refl (refl ∷ 1 ∷ʳ refl ∷ 3 ∷ʳ []) (zero ∷ suc zero ∷ []))
+    ∋ refl
+    where open import Data.List.Relation.Binary.Sublist.Ext
+
+  -- (7) [LIMITATION] does not work well with pattern lambdas
+  refl? : ℕ → Set
+  refl? n = case n ≟ n of λ where
+    (yes p) → ⊤
+    (no ¬p) → ⊥
+
+  -- unquoteDecl refl42 = inlineDecl refl42 (quoteTerm (refl? 42))
+  {-
+  refl42
```

Comment:
This fails because applying pattern-lambdas directly usually fails. If you generate the code
```agda
  refl42
    = case (42 ≟ 42) of
      (λ { (true because ofʸ p) → ⊤ ; (false because ofⁿ ¬p) → ⊥ })

```
it'll work.

This is especially stupid since it appears using `getDefinition` gives you some reduced form of the original definition? So in general, `defineFun n cs >>= getDefinition n` does not return `function cs`...

Reply:
Nice workaround! I've now added a special case to use `case` for single-argument pat-lams and the failing example now works (+ we can use `case` like before in `MOf?` so no need to depend on our generalised `if_then_else`).